ghcide/src/Development/IDE/GHC/Compat.hs

```diff
@@ -310,12 +310,25 @@ corePrepExpr _ = GHC.corePrepExpr
 
 renderMessages :: PsMessages -> (Bag WarnMsg, Bag ErrMsg)
 renderMessages msgs =
+#if MIN_VERSION_ghc(9,6,1)
+  let renderMsgs extractor = (fmap . fmap) GhcPsMessage . getMessages $ extractor msgs
+  in (renderMsgs psWarnings, renderMsgs psErrors)
+#else
   let renderMsgs extractor = (fmap . fmap) renderDiagnosticMessageWithHints . getMessages $ extractor msgs
   in (renderMsgs psWarnings, renderMsgs psErrors)
+#endif
 
+#if MIN_VERSION_ghc(9,6,1)
+pattern PFailedWithErrorMessages :: forall a b. (b -> Bag (MsgEnvelope GhcMessage)) -> ParseResult a
+#elif MIN_VERSION_ghc(9,3,0)
```

Comment:
TODO: make this the fallthrough case, we don't support GHC versions below 9.4 any more.